system/rig_system_test.go

```diff
@@ -40,46 +44,460 @@ func TestNewRigSystem(t *testing.T) {
-					t.Errorf("Expected Mac %d, got %d", model.MacEncode(tt.mac), got.Controller.Mac)
-				}
-				if got.Controller.Name != tt.ControllerName {
-					t.Errorf("Expected Name %s, got %s", tt.ControllerName, got.Controller.Name)
+				return
+			}
+
+			// Check Controller.
+			if mismatch := mismatchingFieldValues(got.Controller, tt.expectedSystem.Controller, "Skey", "Mac", "Name", "Type"); len(mismatch) != 0 {
+				vwanted := reflect.ValueOf(tt.expectedSystem.Controller)
+				vgot := reflect.ValueOf(got.Controller)
+				for _, fieldName := range mismatch {
+					t.Errorf("mismatch in %s field, wanted: %v, got: %v", fieldName, vwanted.FieldByName(fieldName), vgot.FieldByName(fieldName))
 				}
-				for i, variable := range got.Variables {
-					if !reflect.DeepEqual(*variable, *(tt.expectedSystem.Variables[i])) {
-						t.Errorf("Expected Variables %v, got %v", *tt.expectedSystem.Variables[i], *variable)
+				t.FailNow()
+			}
+
+			// Check Variables.
+			if len(got.Variables) > 0 {
+				t.Log(*got.Variables[0])
+			}
+			if len(got.Variables) != len(tt.expectedSystem.Variables) {
+				t.Fatalf("mismatch in number of variables, wanted: %d, got: %d", len(got.Variables), len(tt.expectedSystem.Variables))
+			}
+			for i, gotPtr := range got.Variables {
+				got := *gotPtr
+				wanted := *tt.expectedSystem.Variables[i]
+				if mismatch := mismatchingFieldValues(got, wanted, "Skey", "Name", "Value"); len(mismatch) != 0 {
+					vwanted := reflect.ValueOf(wanted)
+					vgot := reflect.ValueOf(got)
+					for _, fieldName := range mismatch {
+						t.Errorf("mismatch in %s field, wanted: %v, got: %v", fieldName, vwanted.FieldByName(fieldName), vgot.FieldByName(fieldName))
 					}
+					t.FailNow()
 				}
-				if !reflect.DeepEqual(got.Sensors, tt.expectedSystem.Sensors) {
-					t.Errorf("Expected Sensors %v, got %v", tt.expectedSystem.Sensors, got.Sensors)
-				}
-				if !reflect.DeepEqual(got.Actuators, tt.expectedSystem.Actuators) {
-					t.Errorf("Expected Actuators %v, got %v", tt.expectedSystem.Actuators, got.Actuators)
+			}
+
+			// Check Sensors.
+			if len(got.Sensors) != len(tt.expectedSystem.Sensors) {
+				t.Fatalf("mismatch in number of sensors, wanted: %d, got: %d", len(got.Sensors), len(tt.expectedSystem.Sensors))
+			}
+			for i, gotPtr := range got.Sensors {
+				got := *gotPtr
+				wanted := *tt.expectedSystem.Sensors[i]
+				if mismatch := mismatchingFieldValues(got, wanted, "Mac", "Name"); len(mismatch) != 0 {
+					vwanted := reflect.ValueOf(wanted)
+					vgot := reflect.ValueOf(got)
+					for _, fieldName := range mismatch {
+						t.Errorf("mismatch in %s field, wanted: %v, got: %v", fieldName, vwanted.FieldByName(fieldName), vgot.FieldByName(fieldName))
+					}
+					t.FailNow()
 				}
-				if !reflect.DeepEqual(got.Peripherals, tt.expectedSystem.Peripherals) {
-					t.Errorf("Expected Peripherals %v, got %v", tt.expectedSystem.Peripherals, got.Peripherals)
+			}
+
+			// Check Actuators.
```

Comment:
The check code for variables, sensors and actuators is basically the same. Depending on how daring you are, you could try to write a function to do this using generics.

Reply:
I think I'll just leave it like it is. I think it makes it a bit easier to extend or change in the future, and doesn't really make anything worse?

Reply:
It'd be for the sake of DRY (don't repeat yourself). I think common functionality/patterns is better factored out into a single place. Normally this is what make code easier to extend or change (you're only change things in one place instead of multiple).

The function would like similar to this

```
// CompareSlices compares two slices of any type and checks for mismatched field values.
func CompareSlices[T any](t *testing.T, gotSlice, wantSlice []*T, fields ...string) {
	if len(gotSlice) != len(wantSlice) {
		t.Fatalf("mismatch in number of items, wanted: %d, got: %d", len(wantSlice), len(gotSlice))
	}

	for i, gotPtr := range gotSlice {
		got := *gotPtr
		wanted := *wantSlice[i]
		if mismatch := mismatchingFieldValues(got, wanted, fields...); len(mismatch) != 0 {
			vwanted := reflect.ValueOf(wanted)
			vgot := reflect.ValueOf(got)
			for _, fieldName := range mismatch {
				t.Errorf("mismatch in %s field, wanted: %v, got: %v", fieldName, vwanted.FieldByName(fieldName), vgot.FieldByName(fieldName))
			}
			t.FailNow()
		}
	}
}
```

Reply:
yeah ok, that is much better than what I was picturing. I have made the change now.